p2p/transport/webrtc/datachannel.go

```diff
@@ -0,0 +1,341 @@
+		readBuf:        []byte{},
+	}
+
+	channel.SetBufferedAmountLowThreshold(bufferedAmountLowThreshold)
+	channel.OnBufferedAmountLow(func() {
+		result.writeAvailable <- struct{}{}
+	})
+
+	return result
+}
+
+func (d *dataChannel) processControlMessage(msg pb.Message) {
+	d.m.Lock()
+	defer d.m.Unlock()
+	if d.state == stateClosed {
+		return
+	}
+	if msg.Flag == nil {
+		return
+	}
+	switch msg.GetFlag() {
+	case pb.Message_FIN:
+		if d.state == stateWriteClosed {
+			d.Close()
+			return
+		}
+		d.state = stateReadClosed
+	case pb.Message_STOP_SENDING:
+		if d.state == stateReadClosed {
+			d.Close()
+			return
+		}
+		d.state = stateWriteClosed
+	case pb.Message_RESET:
+		d.channel.Close()
+	}
+}
+
+func (d *dataChannel) Read(b []byte) (int, error) {
+	for {
+		select {
+		case <-d.readDeadline.wait():
+			return 0, os.ErrDeadlineExceeded
+		default:
+		}
+
+		d.m.Lock()
+		read := copy(b, d.readBuf)
+		d.readBuf = d.readBuf[read:]
+		d.m.Unlock()
+		if state := d.getState(); len(d.readBuf) == 0 && (state == stateReadClosed || state == stateClosed) {
+			return read, io.EOF
+		}
+		if read > 0 {
+			return read, nil
+		}
+
+		// read until data message
+		var msg pb.Message
+		signal := make(chan struct {
+			error
+		})
+
+		// read in a separate goroutine to enable read deadlines
+		go func() {
+			err := d.reader.ReadMsg(&msg)
+			if err != nil {
+				if err != io.EOF {
+					log.Warnf("error reading from datachannel: %v", err)
+				}
+				signal <- struct {
+					error
+				}{err}
+				return
+			}
+
+			if state := d.getState(); state != stateClosed && state != stateReadClosed && msg.Message != nil {
+				d.m.Lock()
+				d.readBuf = append(d.readBuf, msg.Message...)
+				d.m.Unlock()
+			}
+			d.processControlMessage(msg)
+
+			signal <- struct{ error }{nil}
+
+		}()
+		select {
+		case sig := <-signal:
+			if sig.error != nil {
+				return 0, sig.error
+			}
+		case <-d.readDeadline.wait():
+			return 0, os.ErrDeadlineExceeded
+		}
+
+	}
+}
+
+func (d *dataChannel) Write(b []byte) (int, error) {
+	if s := d.getState(); s == stateWriteClosed || s == stateClosed {
+		return 0, io.ErrClosedPipe
+	}
+
+	var err error
+	var (
+		start     int = 0
+		end           = len(b)
+		chunkSize     = int(maxMessageSize) - protoOverhead - varintOverhead
+		n             = 0
+	)
+
+	for start < len(b) {
+		end = len(b)
+		if start+chunkSize < end {
+			end = start + chunkSize
+		}
+		chunk := b[start:end]
+		n, err = d.partialWrite(chunk)
+		if err != nil {
+			break
+		}
+		start += n
+	}
+	return start, err
+}
+
+func (d *dataChannel) partialWrite(b []byte) (int, error) {
+	if s := d.getState(); s == stateWriteClosed || s == stateClosed {
+		return 0, io.ErrClosedPipe
+	}
+	select {
+	case <-d.writeDeadline.wait():
+		return 0, os.ErrDeadlineExceeded
+	default:
+	}
+	msg := &pb.Message{Message: b}
+	if d.channel.BufferedAmount()+uint64(len(b))+uint64(varintOverhead) > maxBufferedAmount {
+		select {
+		case <-d.writeAvailable:
```

Comment:
This should only block if the latest message being sent is going to exceed the maxBufferedAmount. Any calls to write will add the the buffer which will be cleared as data is sent. If data is sent too fast, it will block because `OnBufferedAmount + len(latest message)` will exceed `maxBufferedAmount`.